// egg-ts-helper/dist/command.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Command = void 0;
const path_1 = __importDefault(require("path"));
const commander_1 = require("commander");
Object.defineProperty(exports, "Command", { enumerable: true, get: function () { return commander_1.Command; } });
const assert_1 = __importDefault(require("assert"));
const core_1 = __importStar(require("./core"));
const utils_1 = require("./utils");
class Commander {
    constructor(options) {
        this.commands = (0, utils_1.loadModules)(path_1.default.resolve(__dirname, './cmd'), true);
        this.tsHelperClazz = options?.tsHelperClazz || core_1.default;
        let version = options?.version;
        if (!version) {
            version = (0, utils_1.getPkgInfo)(path_1.default.dirname(__dirname)).version;
        }
        this.program = new commander_1.Command()
            .version(version, '-v, --version')
            .usage('[commands] [options]')
            .option('-w, --watch', 'Watching files, d.ts would recreated while file changed')
            .option('-c, --cwd [path]', 'Egg application base dir (default: process.cwd)')
            .option('-C, --config [path]', 'Configuration file, The argument can be a file path to a valid JSON/JS configuration file.（default: {cwd}/tshelper.js')
            .option('-f, --framework [name]', 'Egg framework(default: egg)')
            .option('-o, --oneForAll [path]', 'Create a d.ts import all types (default: typings/ets.d.ts)')
            .option('-s, --silent', 'Running without output')
            .option('-i, --ignore [dirs]', 'Ignore watchDirs, your can ignore multiple dirs with comma like: -i controller,service')
            .option('-e, --enabled [dirs]', 'Enable watchDirs, your can enable multiple dirs with comma like: -e proxy,other')
            .option('-E, --extra [json]', 'Extra config, the value should be json string');
    }
    init(argv) {
        const { program, commands } = this;
        let executeCmd;
        // override executeSubCommand to support async subcommand.
        program.addImplicitHelpCommand = () => { };
        program.executeSubCommand = async function (argv, args, unknown) {
            const cwd = this.cwd || core_1.defaultConfig.cwd;
            const command = commands[executeCmd];
            (0, assert_1.default)(command, executeCmd + ' does not exist');
            await command.run(this, { cwd, argv, args: args.filter(item => item !== this), unknown });
        };
        if (!argv.slice(2).length) {
            this.execute();
        }
        else {
            Object.keys(commands).forEach(cmd => {
                const subCommand = commands[cmd];
                const cmdName = subCommand.options ? `${cmd} ${subCommand.options}` : cmd;
                program.command(cmdName, subCommand.description)
                    .action(command => executeCmd = command);
            });
            program.parse(argv);
            if (!executeCmd) {
                this.execute();
            }
        }
    }
    execute() {
        const { program } = this;
        const watchFiles = program.watch;
        const generatorConfig = {};
        program.ignore && program.ignore.split(',').forEach(key => (generatorConfig[key] = false));
        program.enabled && program.enabled.split(',').forEach(key => (generatorConfig[key] = true));
        const tsHelperConfig = {
            cwd: program.cwd || core_1.defaultConfig.cwd,
            framework: program.framework,
            watch: watchFiles,
            generatorConfig,
            configFile: program.config,
            ...(program.extra ? JSON.parse(program.extra) : {}),
        };
        // silent
        if (program.silent) {
            tsHelperConfig.silent = true;
        }
        if ((0, utils_1.checkMaybeIsJsProj)(tsHelperConfig.cwd)) {
            // write jsconfig if the project is wrote by js
            (0, utils_1.writeJsConfig)(tsHelperConfig.cwd);
        }
        // create instance
        const clazz = this.tsHelperClazz;
        const tsHelper = new clazz(tsHelperConfig).build();
        if (program.oneForAll) {
            // create one for all
            tsHelper.createOneForAll(program.oneForAll);
        }
    }
}
exports.default = Commander;
