// egg-ts-helper/dist/scripts/eggInfo.js
"use strict";
/**
 * Getting plugin info in child_process to prevent effecting egg application( splitting scopes ).
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const config_1 = require("../config");
const utils = __importStar(require("../utils"));
const cwd = process.cwd();
const eggInfo = {};
const startTime = Date.now();
if (utils.checkMaybeIsTsProj(cwd)) {
    // only require ts-node in ts project
    const tsconfigPath = path_1.default.resolve(cwd, './tsconfig.json');
    if (fs_1.default.existsSync(tsconfigPath)) {
        require('ts-node').register(utils.readJson5(tsconfigPath));
    }
    else {
        require('ts-node/register');
    }
}
// try to read postinstall script env.ETS_SCRIPT_FRAMEWORK, let egg-bin can auto set the default framework
const framework = (utils.getPkgInfo(cwd).egg || {}).framework || process.env.ETS_SCRIPT_FRAMEWORK || 'egg';
const loader = getLoader(cwd, framework);
if (loader) {
    try {
        loader.loadPlugin();
    }
    catch (e) {
        console.warn('[egg-ts-helper] WARN loader.loadPlugin() error: %s, cwd: %s, framework: %s', e, cwd, framework);
        // do nothing
    }
    // hack loadFile, ignore config file without customLoader for faster booting
    mockFn(loader, 'loadFile', filepath => {
        if (filepath && filepath.substring(filepath.lastIndexOf(path_1.default.sep) + 1).startsWith('config.')) {
            const fileContent = fs_1.default.readFileSync(filepath, 'utf-8');
            if (!fileContent.includes('customLoader'))
                return;
        }
        return true;
    });
    try {
        loader.loadConfig();
    }
    catch (e) {
        console.warn('[egg-ts-helper] WARN loader.loadConfig() error: %s, cwd: %s, framework: %s', e, cwd, framework);
        // do nothing
    }
    eggInfo.plugins = loader.allPlugins;
    eggInfo.config = loader.config;
    eggInfo.eggPaths = loader.eggPaths;
    eggInfo.timing = Date.now() - startTime;
}
utils.writeFileSync(config_1.eggInfoPath, JSON.stringify(eggInfo));
/* istanbul ignore next */
function noop() { }
function mockFn(obj, name, fn) {
    const oldFn = obj[name];
    obj[name] = (...args) => {
        const result = fn.apply(obj, args);
        if (result)
            return oldFn.apply(obj, args);
    };
}
function getLoader(baseDir, framework) {
    let frameworkPath = '';
    try {
        frameworkPath = require.resolve(framework, { paths: [baseDir] });
    }
    catch (_) {
        // ignore error
    }
    if (!frameworkPath) {
        frameworkPath = path_1.default.join(baseDir, 'node_modules', framework);
    }
    const eggCore = findEggCore(baseDir, frameworkPath);
    if (!eggCore) {
        console.warn('[egg-ts-helper] WARN cannot find egg core from frameworkPath: %s', frameworkPath);
        return;
    }
    const EggLoader = eggCore.EggLoader;
    const egg = utils.requireFile(frameworkPath) || utils.requireFile(framework);
    if (!egg || !egg.Application || !EggLoader)
        return;
    process.env.EGG_SERVER_ENV = 'local';
    return new EggLoader({
        baseDir,
        logger: {
            debug: noop,
            info: noop,
            warn: noop,
            error: noop,
        },
        app: Object.create(egg.Application.prototype),
    });
}
function findEggCore(baseDir, frameworkPath) {
    let eggCorePath = '';
    try {
        eggCorePath = require.resolve('egg-core', { paths: [frameworkPath] });
    }
    catch (_) {
        // ignore error
    }
    if (!eggCorePath) {
        eggCorePath = path_1.default.join(baseDir, 'node_modules/egg-core');
        if (!fs_1.default.existsSync(eggCorePath)) {
            eggCorePath = path_1.default.join(frameworkPath, 'node_modules/egg-core');
        }
    }
    // try to load egg-core in cwd
    const eggCore = utils.requireFile(eggCorePath);
    if (!eggCore) {
        // try to resolve egg-core
        return utils.requireFile('egg-core');
    }
    return eggCore;
}
