// @eggjs/utils/dist/commonjs/plugin.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPlugins = getPlugins;
exports.getLoadUnits = getLoadUnits;
exports.getConfig = getConfig;
exports.getLoader = getLoader;
exports.findEggCore = findEggCore;
const node_util_1 = require("node:util");
const node_path_1 = __importDefault(require("node:path"));
const node_assert_1 = __importDefault(require("node:assert"));
const node_os_1 = __importDefault(require("node:os"));
const promises_1 = require("node:fs/promises");
const import_js_1 = require("./import.js");
const debug = (0, node_util_1.debuglog)('@eggjs/utils/plugin');
const tmpDir = node_os_1.default.tmpdir();
function noop() { }
const logger = {
    debug: noop,
    info: noop,
    warn: noop,
    error: noop,
};
/**
 * @see https://github.com/eggjs/egg-core/blob/2920f6eade07959d25f5c4f96b154d3fbae877db/lib/loader/mixin/plugin.js#L203
 */
async function getPlugins(options) {
    const loader = await getLoader(options);
    await loader.loadPlugin();
    return loader.allPlugins;
}
/**
 * @see https://github.com/eggjs/egg-core/blob/2920f6eade07959d25f5c4f96b154d3fbae877db/lib/loader/egg_loader.js#L348
 */
async function getLoadUnits(options) {
    const loader = await getLoader(options);
    await loader.loadPlugin();
    return loader.getLoadUnits();
}
async function getConfig(options) {
    const loader = await getLoader(options);
    await loader.loadPlugin();
    await loader.loadConfig();
    return loader.config;
}
async function exists(filepath) {
    try {
        await (0, promises_1.stat)(filepath);
        return true;
    }
    catch {
        return false;
    }
}
async function getLoader(options) {
    (0, node_assert_1.default)(options.framework, 'framework is required');
    (0, node_assert_1.default)(await exists(options.framework), `${options.framework} should exist`);
    if (!(options.baseDir && await exists(options.baseDir))) {
        options.baseDir = node_path_1.default.join(tmpDir, 'egg_utils', `${Date.now()}`, 'tmp_app');
        await (0, promises_1.mkdir)(options.baseDir, { recursive: true });
        await (0, promises_1.writeFile)(node_path_1.default.join(options.baseDir, 'package.json'), JSON.stringify({
            name: 'tmp_app',
        }));
        debug('[getLoader] create baseDir: %o', options.baseDir);
    }
    const { EggCore, EggLoader } = await findEggCore(options);
    const mod = await (0, import_js_1.importModule)(options.framework);
    const Application = mod.Application ?? mod.default?.Application;
    (0, node_assert_1.default)(Application, `Application not export on ${options.framework}`);
    if (options.env) {
        process.env.EGG_SERVER_ENV = options.env;
    }
    return new EggLoader({
        baseDir: options.baseDir,
        logger,
        app: Object.create(Application.prototype),
        EggCoreClass: EggCore,
    });
}
async function findEggCore(options) {
    const baseDirRealpath = await (0, promises_1.realpath)(options.baseDir);
    const frameworkRealpath = await (0, promises_1.realpath)(options.framework);
    const paths = [frameworkRealpath, baseDirRealpath];
    // custom framework => egg => @eggjs/core
    try {
        const { EggCore, EggLoader } = await (0, import_js_1.importModule)('egg', { paths });
        if (EggLoader) {
            return { EggCore, EggLoader };
        }
    }
    catch (err) {
        debug('[findEggCore] import "egg" from paths:%o error: %o', paths, err);
    }
    // egg-core 在 6.2.3 版本中更名为 @eggjs/core，为兼容老版本，支持同时查找两个包，优先使用新名字
    const names = ['@eggjs/core', 'egg-core'];
    for (const name of names) {
        try {
            const { EggCore, EggLoader } = await (0, import_js_1.importModule)(name, { paths });
            if (EggLoader) {
                return { EggCore, EggLoader };
            }
        }
        catch (err) {
            debug('[findEggCore] import "%s" from paths:%o error: %o', name, paths, err);
        }
        try {
            const { EggCore, EggLoader } = await (0, import_js_1.importModule)(name);
            if (EggLoader) {
                return { EggCore, EggLoader };
            }
        }
        catch (err) {
            debug('[findEggCore] import "%s" error: %o', name, err);
        }
        let eggCorePath = node_path_1.default.join(options.baseDir, `node_modules/${name}`);
        if (!(await exists(eggCorePath))) {
            eggCorePath = node_path_1.default.join(options.framework, `node_modules/${name}`);
        }
        if (await exists(eggCorePath)) {
            return await (0, import_js_1.importModule)(eggCorePath);
        }
    }
    (0, node_assert_1.default)(false, `Can't find ${names.join(' or ')} from ${options.baseDir} and ${options.framework}`);
}
